sources/api/early-boot-config/src/main.rs

```diff
@@ -94,48 +95,144 @@ fn parse_args(args: env::Args) -> Args {
-    SimpleLogger::init(args.log_level, LogConfig::default()).context(error::LoggerSnafu)?;
+/// Check that a user data provider succeeded and forward its logs
+fn check_provider_status<P>(provider: P, output: &process::Output) -> Result<()>
+where
+    P: AsRef<Path>,
+{
+    let provider = provider.as_ref();
+    // Regardless of provider status, log its output
+    let provider_name = provider
+        .file_name()
+        .unwrap_or(provider.as_os_str())
+        .to_string_lossy();
+    let provider_logs = String::from_utf8_lossy(&output.stderr);
+    for line in provider_logs.lines() {
+        info!("Provider '{}': {}", provider_name, line);
+    }
 
-    info!("early-boot-config started");
+    ensure!(
+        output.status.success(),
+        error::ProviderFailureSnafu {
+            provider: &provider,
+        }
+    );
 
-    info!("Retrieving platform-specific data");
+    Ok(())
+}
+
+/// Submit user data to the API
+async fn submit_user_data<S>(socket_path: S, user_data: serde_json::Value) -> Result<()>
+where
+    S: AsRef<str>,
+{
+    let socket_path = socket_path.as_ref();
     let uri = &format!(
         "{}?tx={}",
         constants::API_SETTINGS_URI,
         constants::LAUNCH_TRANSACTION
     );
     let method = "PATCH";
-    for settings_json in Platform
-        .platform_data()
-        .await
-        .context(error::ProviderSnafu)?
-    {
-        // Don't send an empty request to the API
-        if settings_json.json.is_empty() {
-            warn!("{} was empty", settings_json.desc);
-            continue;
+    trace!("Request body: {}", user_data);
+
+    let (code, response_body) =
+        apiclient::raw_request(socket_path, uri, method, Some(user_data.to_string()))
+            .await
+            .context(error::APIRequestSnafu { method, uri })?;
+
+    ensure!(
+        code.is_success(),
+        error::ResponseSnafu {
+            method,
+            uri,
+            code,
+            response_body,
         }
+    );
+    Ok(())
+}
 
-        info!("Sending {} to API", settings_json.desc);
-        trace!("Request body: {}", settings_json.json);
-        let (code, response_body) =
-            apiclient::raw_request(&args.socket_path, uri, method, Some(settings_json.json))
+async fn run() -> Result<()> {
+    // Parse and store the args passed to the program
+    let args = parse_args(env::args());
+
+    env_logger::Builder::new()
+        .filter_level(args.log_level)
+        .format_module_path(false)
+        .target(Target::Stdout)
+        .write_style(WriteStyle::Never)
+        .init();
+
+    info!("early-boot-config started");
+
+    info!("Gathering user data providers");
+    let mut threads = Vec::new();
+    let providers = gather_providers()?;
+    for provider in providers {
+        threads.push((
+            provider.clone(),
+            tokio::spawn(async move { run_provider(&args.log_level, &provider).await }),
+        ));
```

Comment:
Might be better to use [FuturesOrdered](https://docs.rs/futures/latest/futures/stream/struct.FuturesOrdered.html) instead:
```
let mut s = stream::FuturesOrdered::new();
let providers = gather_providers()?;
for provider in providers {
    s.push_back(...);
}

while Some(result) = s.next().await? {
    // ...
}
```

Reply:
Been trying a few approaches to use this, but it keeps getting messy because we need the provider name for logging and errors when we join the threads. I tried having the thread return `(provider_name, result)` as well as creating a custom struct that impls `Future`, but I think both ended up less readable than the `Vec<(provider, JoinHandle)>` approach -- the return from `.next()` always needed to be converted into multiple other types, and

```
while let Some(Ok(provider, result)) = s.next().await {
...
}
```

doesn't handle errors correctly.

I think we should stick with the current solution unless you feel there's a compelling performance reason to use `FuturesOrdered`.

Reply:
> I think we should stick with the current solution unless you feel there's a compelling performance reason to use FuturesOrdered.

Sounds good!